config/policy.go

```diff
@@ -568,35 +637,171 @@ func (p *Policy) Validate() error {
+	pr.Response = pr.pooledFields.response
+	pr.Timeout = pr.pooledFields.timeout
+	pr.IdleTimeout = pr.pooledFields.idleTimeout
+}
+
+func (pr *pooledRoute) reset() {
+	// Reset to clear any pointers that may be aliasing other fields in the
+	// original Policy, to avoid extending its lifetime
+	pr.Reset()
+	pr.prepare() // copy back the fields we own
+	clear(pr.To)
+	pr.To = pr.To[:0]
+	clear(pr.LoadBalancingWeights)
+	pr.LoadBalancingWeights = pr.LoadBalancingWeights[:0]
+	pr.Redirect.Reset() // contains possibly aliased pointers
+	pr.Response.Reset()
+}
+
+// pooledFields contains a subset of fields that are owned by the pooled
+// Route objects, and reused along with the Route itself.
+// The fields here are either slices for commonly used fields that are stored
+// in a different way in Policy, or proto message types that cannot be directly
+// aliased from the Policy and would need to be heap-allocated each time.
+type pooledFields struct {
+	nameBuffer           []byte
+	to                   []string
+	loadBalancingWeights []uint32
+	redirect             *configpb.RouteRedirect
+	response             *configpb.RouteDirectResponse
+	timeout              *durationpb.Duration
+	idleTimeout          *durationpb.Duration
+}
+
+var checksumBufferPool bytebufferpool.Pool
+
+var marshalOpts = proto.MarshalOptions{
+	Deterministic: true,
+}
+
+var (
+	setRequestHeadersIv  = xxhash.Sum64String("Policy.SetRequestHeaders")
+	setResponseHeadersIv = xxhash.Sum64String("Policy.SetResponseHeaders")
+)
+
 // Checksum returns the xxhash hash for the policy.
 func (p *Policy) Checksum() uint64 {
-	return hashutil.MustHash(p)
+	return p.ChecksumWithID(p.MustRouteID())
 }
 
-// RouteID returns a unique identifier for a route
-func (p *Policy) RouteID() (uint64, error) {
-	id := routeID{
-		From:   p.From,
-		Prefix: p.Prefix,
-		Path:   p.Path,
-		Regex:  p.Regex,
-	}
+func (p *Policy) ChecksumWithID(routeID uint64) uint64 {
+	pr := policyPool.Get().(*pooledRoute)
+	pr.prepare()
+	p.CopyToProto(pr.Route)
+	pr.unsafeSetName(routeID)
+	var setReqHeaders, setRespHeaders map[string]string
+	setReqHeaders, pr.SetRequestHeaders = pr.SetRequestHeaders, nil
+	setRespHeaders, pr.SetResponseHeaders = pr.SetResponseHeaders, nil
+	buf := checksumBufferPool.Get()
+	buf.B, _ = marshalOpts.MarshalAppend(buf.B, pr)
+	if setReqHeaders != nil {
+		buf.B = binary.BigEndian.AppendUint64(buf.B, hashutil.MapHash(setRequestHeadersIv, setReqHeaders))
+	}
+	if setRespHeaders != nil {
+		buf.B = binary.BigEndian.AppendUint64(buf.B, hashutil.MapHash(setResponseHeadersIv, setRespHeaders))
+	}
+	sum := xxhash.Sum64(buf.B)
+	checksumBufferPool.Put(buf)
+	pr.reset()
+	policyPool.Put(pr)
+	return sum
+}
 
+// RouteID returns a unique identifier for a route.
+// Only the fields From, To, Redirect, and Response are used to compute the ID.
+func (p *Policy) RouteID() (uint64, error) {
+	// this function is in the hot path, try not to allocate too much memory here
+	hash := xxhash.New()
```

Comment:
I had an idea. What about creating a new `Digest` type in `hashutil` to wrap `xxhash.Digest` so we can add these additional methods:

```go
package hashutil

import (
	"encoding/binary"

	"github.com/cespare/xxhash/v2"
)

// Digest is a hash digest computed via xxhash.Digest.
type Digest struct {
	xxhash.Digest
}

// NewDigest creates a new Digest.
func NewDigest() *Digest {
	return &Digest{
		Digest: *xxhash.New(),
	}
}

// WriteBool writes a bool.
func (hash *Digest) WriteBool(b bool) {
	if b {
		hash.Write([]byte{1})
	} else {
		hash.Write([]byte{0})
	}
}

// WriteUint32 writes a uint32.
func (hash *Digest) WriteUint32(t uint32) {
	var buf [4]byte
	binary.LittleEndian.PutUint32(buf[:], t)
	hash.Write(buf[:])
}

// WriteInt32 writes an int32.
func (hash *Digest) WriteInt32(t int32) {
	var buf [4]byte
	binary.LittleEndian.PutUint32(buf[:], uint32(t))
	hash.Write(buf[:])
}
```

Maybe could add a `WriteNullableString` (or similarly named) method to cut down on all the ifs too.

Reply:
Incidentally, for those nullable fields, we might need to write a `0` byte or something... otherwise how to do we tell the difference between:

```
{ "httpsRedirect": "VALUE" }
```

and

```
{ "schemeRedirect": "VALUE" }
```

wouldn't they hash to the same thing?

Reply:
Good idea; and also yes you are correct, this is missing some null bytes (also unit tests)